Support multiple response header values in raw web actions

bbrowning opened the pull request:
This change allows multiple response header values to be set in raw
web actions by using a JSON array as the header value. For example:

```
function main() {
  return {
    headers: {
      "Set-Cookie": ["a=b", "c=d"]
    },
    code: 200
  }
}
```

I have signed the Apache CLA as "Benjamin M. Browning".

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -216,11 +216,8 @@ protected[core] object WhiskWebActionsApi extends Directives {
-                    case _                 => throw new Throwable("Invalid header")
-                }.toList
+                    case (k, v) => headersFromJson(k, v)
+                }.flatten.toList
```
You should be able to use `flatMap` above to omit the extra `flatten`

markusthoemmes commented on tests/src/test/scala/whisk/core/controller/test/WebActionsApiTests.scala:
```diff
@@ -1127,6 +1127,27 @@ trait WebActionsApiTests extends ControllerTestCommon with BeforeAndAfterEach wi
+                            headers.length shouldBe 2
+                            headers(0).toString shouldBe "Set-Cookie: a=b"
+                            headers(1).toString shouldBe "Set-Cookie: c=d; Path = /"
+                        }
```
Using the extensive matchers of ScalaTest, we can write this in a more speaking way and even preserve great error messages in case a test fails. We also should preserve typesafety here and not rely on the string representation of the headers:

Assuming you want to keep the order stable (which I think is not necessary here) you could write:

```scala
headers shouldBe List(RawHeader("Set-Cookie", "a=b"), RawHeader("Set-Cookie", "c=d; Path = /"))
```

This code would fail with appropriate error messages if the list doesn't comply to the test, and automatically print both lists to be checked thus removing the need for `withClue` and a specific length check.

As I think we shouldn't check for a specific order here, we could use `contains allOf` like so:

```scala
headers should contain allOf (
    RawHeader("Set-Cookie", "a=b"),
    RawHeader("Set-Cookie", "c=d; Path = /")
)
```

Externalizing the cookie values and the headername (maybe even strongly typing them with `scala.http.HttpHeaders.Set-Cookie` and `scala.http.HttpCookie`) is valuable here as well.

markusthoemmes commented:
Great contribution, thank you :+1:

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/WebActions.scala:
```diff
@@ -251,6 +248,14 @@ protected[core] object WhiskWebActionsApi extends Directives {
+        case JsString(v)  => Seq(RawHeader(k, v))
+        case JsBoolean(v) => Seq(RawHeader(k, v.toString))
+        case JsNumber(v)  => Seq(RawHeader(k, v.toString))
+        case JsArray(v)   => v.map(headersFromJson(k, _:JsValue)).flatten
```
Same comment on `flatMap`. In general, nesting a `_` can become tricky. I'd prefer fat-arrow notation in this case, the non inferred type seems confusing.

```scala
case JsArray(v)   => v.flatMap(inner => headersFromJson(k, inner))
```

bbrowning commented:
Thanks for the pointers! I pushed another commit addressing the issues raised so far. Normally I'd squash/rebase and do a force push, but wasn't sure how the project preferred to handle that during code reviews and before merging.

rabbah commented:
Thanks for the contribution and for not squashing the revision! Indeed it helps with reviews not to constantly squash.

Usually the committers can rebase, squash and update the commit message as the PR is merged.

markusthoemmes commented:
PG1 1908 🔵

markusthoemmes commented:
@bbrowning congratulations to your first contribution and welcome to the OpenWhisk gang 🎉
